aclk/aclk_query.c

```diff
@@ -0,0 +1,581 @@
+            ACLK_SHARED_STATE_LOCK;
+            meta_state = aclk_shared_state.metadata_submitted;
+            aclk_shared_state.metadata_submitted = ACLK_METADATA_SENT;
+            ACLK_SHARED_STATE_UNLOCK;
+            aclk_send_metadata(meta_state);
+            break;
+
+        case ACLK_CMD_CHART:
+            debug(D_ACLK, "EXECUTING a chart update command");
+            aclk_send_single_chart(this_query->data, this_query->query);
+            break;
+
+        case ACLK_CMD_CHARTDEL:
+            debug(D_ACLK, "EXECUTING a chart delete command");
+            //TODO: This send the info metadata for now
+            aclk_send_info_metadata(ACLK_METADATA_SENT);
+            break;
+
+        case ACLK_CMD_ALARM:
+            debug(D_ACLK, "EXECUTING an alarm update command");
+            aclk_send_message(this_query->topic, this_query->query, this_query->msg_id);
+            break;
+
+        case ACLK_CMD_CLOUD:
+            t = now_monotonic_high_precision_usec();
+            debug(D_ACLK, "EXECUTING a cloud command");
+            aclk_execute_query(this_query);
+            t = now_monotonic_high_precision_usec() - t;
+            break;
+
+        default:
+            break;
+    }
+    debug(D_ACLK, "Query #%ld (%s) done", query_count, this_query->topic);
+
+    if (aclk_stats_enabled) {
+        ACLK_STATS_LOCK;
+        aclk_metrics_per_sample.queries_dispatched++;
+        aclk_metrics_per_sample.queries_per_thread[t_idx]++;
+        if(this_query->cmd == ACLK_CMD_CLOUD) {
+            aclk_metrics_per_sample.cloud_q_process_total += t;
+            aclk_metrics_per_sample.cloud_q_process_count++;
+            if(aclk_metrics_per_sample.cloud_q_process_max < t)
+                aclk_metrics_per_sample.cloud_q_process_max = t;
+        }
+        ACLK_STATS_UNLOCK;
+    }
+
+    aclk_query_free(this_query);
+
+    return 1;
+}
+
+void aclk_query_threads_cleanup(struct aclk_query_threads *query_threads)
+{
+    if (query_threads) {
+        for (int i = 0; i < query_threads->count; i++) {
+            netdata_thread_join(*(query_threads->thread_list[i]->thread), NULL);
+            freez(query_threads->thread_list[i]->thread);
+            freez(query_threads->thread_list[i]);
+        }
+    }
+
+    struct aclk_query *this_query;
+
+    do {
+        this_query = aclk_queue_pop();
+        aclk_query_free(this_query);
+    } while (this_query);
+}
+
+void aclk_query_threads_start(struct aclk_query_threads *query_threads)
+{
+    info("Starting %d query threads.", query_threads->count);
+
+#if ACLK_MAX_QUERY_THREADS > 10
+#error "You seem to have increased ACLK_MAX_QUERY_THREADS above 10. If you want to do that you need to update following thread name generation."
+#endif
+
+    char thread_name[strlen(ACLK_THREAD_NAME) + 3];
+    strcpy(thread_name, ACLK_THREAD_NAME);
+    thread_name[strlen(ACLK_THREAD_NAME)] = '_';
+    thread_name[strlen(ACLK_THREAD_NAME) + 2] = 0;
+
```

Comment:
Why not just build it with sprintf ? 
If someone defines over 10 threads they will kill compilation and then they need to try to fix this.

Reply:
when I was doing this I was thinking 640kb (10 threads) is enough for everyone... this is the only reason... In fact I think the stats thread uses sprintf for the same purpose (when generating dimension names)